Bug fix, description added, function is run

SjoerdGlaser opened the pull request:
Three changes have been made:

A bug is fixed because the analyses could not be performed with 2 items. Now it only runs when at least three items are present.

A description of the function has been added.

A call to run the function has been added.

Futhermore, files of the new project have been removed, so it only contains an update of the old analysis script.

mvdriel commented:
@SjoerdGlaser Could you please remove the 6 hidden files (starting with a dot).

mvdriel commented on app/Lib/Rscripts/analyse.R:
```diff
@@ -103,3 +123,5 @@ Analyse <- function(key, input.answers, number.answeroptions) {
   }
 }
+
+Analyse(key, input.answers, number.answeroptions)
```
Is this a mistake?

mvdriel commented on app/Lib/Rscripts/analyse.R:
```diff
@@ -2,11 +2,31 @@
 
-  if (number.questions > 1 & number.students > 1) {
-    # Correct/Incorrect Matrix
+  if (number.questions > 2 & number.students > 1) { # Do the analysis only if there are at least 3 items and 2 students.
```
Check the Google's R Style Guide: https://google.github.io/styleguide/Rguide.xml#comments

mvdriel commented on app/Lib/Rscripts/analyse.R:
```diff
@@ -2,11 +2,31 @@
+  # Calculates standard psychometric properties of an exam, notably the percentage correct and the item rest correlation (IRC) per item and 
+  # per answer option and the cronbach's alpha for the whole exam.
+  #
+  # Arguments:
```
Check the Google's R Style Guide: https://google.github.io/styleguide/Rguide.xml#functiondefinition

mvdriel replied:
How about the minimal (and maximum) sizes of the arguments?

SjoerdGlaser commented:
The files have been removed. Size requirements for the arguments have been added. I think I fixed the comment with the Google's R style guide, but I'm not sure what you meant.

I added how to run the file because I thought you requested it by mail, but I removed it again.

mvdriel commented on app/Lib/Rscripts/analyse.R:
```diff
@@ -2,11 +2,45 @@
+  # percentage correct and the item rest correlation (IRC) per item and
+  # per answer option and the cronbach's alpha for the whole exam.
+  #
+  # Arguments:
```
"(...) a list of the function's arguments, denoted by Args: (...)" see: https://google.github.io/styleguide/Rguide.xml#functiondocumentation

SjoerdGlaser replied:
Fixed

mvdriel replied:
@SjoerdGlaser I don't see any new commits. Is there something wrong?

SjoerdGlaser replied:
I think I forgot to sync. I synced it just now. I am not sure, but I think I do not need to do a new pull request as the changes seem to have been added to the current pull request as the appear below.

mvdriel commented on app/Lib/Rscripts/analyse.R:
```diff
@@ -2,11 +2,45 @@
+	#        Should be at least of length 3 (3 items), there is no maximum length.
+  #   input.answers: Ungraded matrix of answers. input.answers[i,j] is 
+	#                  the answer of student (i) to item (j). Should consist of
+	#                  at least 3 rows (items) and 2 columns (items).
```
Items and items?

SjoerdGlaser replied:
Fixed

mvdriel commented on app/Lib/Rscripts/analyse.R:
```diff
@@ -2,11 +2,45 @@
+  #   input.answers: Ungraded matrix of answers. input.answers[i,j] is 
+	#                  the answer of student (i) to item (j). Should consist of
+	#                  at least 3 rows (items) and 2 columns (items).
+  #                  Number of collumns should be equal to the length of key and 
```
collumns Typo?

SjoerdGlaser replied:
Yep, fixed

mvdriel commented on app/Lib/Rscripts/analyse.R:
```diff
@@ -2,11 +2,45 @@
+  #
+  # Arguments:
+  #   key: Matrix of 0's and 1's. key[i,j] implies wether answer option i 
+	#        to item j is right (1) or wrong (0). If a row (item) consists of 
```
"(...) When indenting your code, use two spaces.  Never use tabs or mix tabs and spaces. (...)" see: https://google.github.io/styleguide/Rguide.xml#indentation

SjoerdGlaser replied:
Now two spaces are used consistently.

mvdriel commented:
@SjoerdGlaser Thanks. I would like to make use of a code style checker / linter for the R code, but it seems that [rlint] (https://code.google.com/p/google-rlint/) was never released. [lintr] (https://github.com/jimhester/lintr) is an option, but is based on [Hadley Wickham's R Style Guide] (http://r-pkgs.had.co.nz/style.html). Do you have a suggestion?

tersmitten commented:
@mvdriel You can ask Jonathan Klaiber and @Yennick about R linters / code checkers

mvdriel commented:
@tersmitten Thanks!

SjoerdGlaser commented:
@mvdriel I tried the lintr package. Updated one = to a <- and removed a lot of spaces at the end of lines. Furthermore, it suggested to change the names of all variables by using _'s instead of .'s. I did not change this because it is incongruent with the Google R guide, which we have been using.

mvdriel commented:
@SjoerdGlaser Thanks, it looks smooth! :)

mvdriel commented:
@SjoerdGlaser Let's discuss your suggestion (_ ==> .) with @ShKlinkenberg You have my support

mvdriel commented:
@SjoerdGlaser Please, update your branch and resolve conflicts to make merge possible.

SjoerdGlaser commented:
@mvdriel updated with your branch and resolved the conflicts.

mvdriel commented:
I think something went wrong resolving the conflict. @SjoerdGlaser Please check the diff

SjoerdGlaser commented:
@mvdriel Sorry, I do not really understand the issue. My github program says this branch is up-to-date with your master branch. Besides that, if I check the error above, it seems the error is caused by a conflict in the analyse.R file. But I actually want to actually this file, so those two should be a conflict, right?

mvdriel commented:
@SjoerdGlaser Please, review the [diff](https://github.com/nlware/qdnatool/pull/38/files) you created. There is an unresolved conflict in it.

SjoerdGlaser commented:
@mvdriel Is it fixed now?

mvdriel commented:
@SjoerdGlaser Conflict is resolved, but you (accidentally) removed some comments. Please, review your [last commit](https://github.com/SjoerdGlaser/qdnatool/commit/5781450d07a16aae7d2e10601676dff4044cc4bd)

SjoerdGlaser commented:
@mvdriel fixed

mvdriel commented:
@SjoerdGlaser Still some of your previous changes/updates are lost. Please, check the [diff] (https://github.com/SjoerdGlaser/qdnatool/compare/1dfb9abe7b4c8c612f8fb869b3ed37fde1294392...05dc9b0b3552d8c0f51993905c330710a5bf4d36) between your last commit and the commit before resolving the conflict.

mvdriel commented:
@SjoerdGlaser Please, remove the reintroduced spaces. See [diff](https://github.com/SjoerdGlaser/qdnatool/compare/1dfb9abe7b4c8c612f8fb869b3ed37fde1294392...22a671c74c812ec3dcf03d66233becac312a771b)

mvdriel commented on app/Lib/Rscripts/analyse.R:
```diff
@@ -2,12 +2,46 @@
+  #
+  # Args:
+  #   key: Matrix of 0's and 1's. key[i,j] implies wether answer option i
+  #        to item j is right (1) or wrong (0). If a row (item) consists of
```
@SjoerdGlaser Are you sure the items are in the rows?

mvdriel commented on app/Lib/Rscripts/analyse.R:
```diff
@@ -2,12 +2,46 @@
+  #   input.answers: Ungraded matrix of answers. input.answers[i,j] is
+  #                  the answer of student (i) to item (j). Should consist of
+  #                  at least 3 rows (items) and 2 columns (students).
+  #                  Number of columns should be equal to the length of key and
```
@SjoerdGlaser What is the length of a matrix?

mvdriel commented on app/Lib/Rscripts/analyse.R:
```diff
@@ -2,12 +2,47 @@
+  #                  columns of key and length of number.answeroptions
+  #                  number.answeroptions. There is no maximum.
+  #   number.answersoptions: Vector with number of answer options per item,
+  #                          length should be equal to length of key and number
```
@SjoerdGlaser What is the length of a matrix?

mvdriel commented on app/Lib/Rscripts/analyse.R:
```diff
@@ -2,12 +2,47 @@
+  #                  at least 3 columns (items) and 2 rows (students).
+  #                  Number of columns should be equal to the number of
+  #                  columns of key and length of number.answeroptions
+  #                  number.answeroptions. There is no maximum.
```
@SjoerdGlaser Typo?

SjoerdGlaser commented:
Removed weird white space from the comments

Removed extra curly bracket to make function work again.

mvdriel commented on app/Lib/Rscripts/analyse.R:
```diff
@@ -55,7 +55,7 @@ Analyse <- function(key, input.answers, number.answeroptions) {
           input.correct[i, j] <- input.answers[i, j]
         } else {
-          if(any(input.answers[i, j] == which(key[, j] == 1))) {
+          if(any(input.answers[i, j] == which(key[, j] == 1))) 
```
Please, check the style guide about spacing (https://google.github.io/styleguide/Rguide.xml#spacing) and curly braces (https://google.github.io/styleguide/Rguide.xml#curlybraces).

mvdriel commented on app/Lib/Rscripts/analyse.R:
```diff
@@ -2,25 +2,67 @@
           input.correct[i, j] <- input.answers[i, j]
-        } else if(any(input.answers[i, j] == which(key[, j] == 1))) {
-          input.correct[i, j] <- 1
+        } else {
```
The current construction is easier to read. @SjoerdGlaser What is wrong with the "else if"?

SjoerdGlaser replied:
The style guide says "else" should always be surrounded by brackets: https://google.github.io/styleguide/Rguide.xml#else

mvdriel replied:
I don't think it is a problem to use the "else if" construction, because in that situation the else statement is still surrounded on the same line by curly braces.

SjoerdGlaser replied:
I guess it's a matter of interpretation of the guide. I read it as: else should only be used as "} else {" on a separate line. But your suggestion technically follows the guide as well and I agree it is more readable. I removed this change.

mvdriel replied:
@SjoerdGlaser Thanks!

mvdriel commented on app/Lib/Rscripts/analyse.R:
```diff
@@ -94,9 +136,14 @@ Analyse <- function(key, input.answers, number.answeroptions) {
+
+    # Computes Cronbach's Alpha for overall test
+    # If there is no variance in the total score returns -100
+    if(!all(rowSums(input.correct) == rowSums(input.correct)[1])){
```
Please, check the style guide about spacing (https://google.github.io/styleguide/Rguide.xml#spacing).

SjoerdGlaser replied:
I assume you mean the missing space before the curly bracket at the end. I fixed that.

SjoerdGlaser replied:
I did, but we do not follow it on two points, resulting in so many warnings that I do not bother anymore. The first is that the linter package follows a different guide which forbids '.' in variable names, while the google guide encourages it. The second is that our variable names are too long. I think we agreed a while back on not changing the names for consistency. I kinda lost sight of the forest (other style errors) behind the trees (these two erros). However, I did use it again and I hope it is okay now.

mvdriel replied:
@SjoerdGlaser Thanks!

mvdriel commented on app/Lib/Rscripts/analyse.R:
```diff
@@ -2,25 +2,67 @@
+  # per answer option and the cronbach's alpha for the whole exam.
+  #
+  # Args:
+  #   key: Matrix of 0's and 1's. key[i,j] implies wether answer option i
```
"wether" Typo?

SjoerdGlaser replied:
Fixed

mvdriel commented on app/Lib/Rscripts/analyse.R:
```diff
@@ -94,9 +136,14 @@ Analyse <- function(key, input.answers, number.answeroptions) {
+    if(!all(rowSums(input.correct) == rowSums(input.correct)[1])){
+    	cronbach <- round(cronbach(input.correct)$alpha, digits = 3)
+    } else {
+    	cronbach <- -100
```
@SjoerdGlaser Is it possible to make this a value in the range from -9.999 until 9.999? In the current database schema it is not possible to save a value out of this range.

SjoerdGlaser replied:
Changed to -9. The reason for -100 was that it should actually be -Inf, Changed it to the smallest possible value.
